typing/typemod.ml

```diff
@@ -503,305 +510,292 @@ let merge_constraint_aux initial_env loc sg lid constr : merge_result =
-        end
-    | Sig_type(id, sig_decl, rs, priv), [s], With_type_package cty
-      when Ident.name id = s ->
-        begin match sig_decl.type_manifest with
-        | None -> ()
-        | Some ty ->
-          raise (Error(loc, outer_sig_env, With_package_manifest (lid.txt, ty)))
-        end;
-        Env.mark_type_used sig_decl.type_uid;
-        let tdecl =
-          Typedecl.transl_package_constraint ~loc outer_sig_env cty.ctyp_type
-        in
-        check_type_decl outer_sig_env sg_for_env loc id None tdecl sig_decl;
-        let tdecl = { tdecl with type_manifest = None } in
-        return ~ghosts ~replace_by:(Some(Sig_type(id, tdecl, rs, priv)))
-          (Pident id, No_TypedTree)
-    | Sig_modtype(id, mtd, priv), [s],
-      (With_modtype mty | With_modtypesubst mty)
-      when Ident.name id = s ->
-        let new_item = patch_modtype_item id mtd priv mty.mty_type in
-        let constr_tt =
-          if not destructive_substitution then
-            (Twith_modtype mty)
-          else
-            (Twith_modtypesubst mty)
-        in
-        return ~replace_by:new_item
-          (Pident id, Built_TypedTree {lid; constr=constr_tt})
-    | Sig_modtype(id, mtd, priv), [s],
-      (Approx_with_modtype mty | Approx_with_modtypesubst mty)
-      when Ident.name id = s ->
-        let new_item = patch_modtype_item id mtd priv mty in
-        return ~replace_by:new_item (Pident id, No_TypedTree)
-    | Sig_module(id, pres, md, rs, priv), [s],
-      With_module {lid=lid'; md=md'; path; remove_aliases}
-      when Ident.name id = s ->
-        let sig_env = Env.add_signature sg_for_env outer_sig_env in
-        let mty = md'.md_type in
-        let mty = Mtype.scrape_for_type_of ~remove_aliases sig_env mty in
-        let md'' = { md' with md_type = mty } in
-        let newmd = Mtype.strengthen_decl ~aliasable:false sig_env md'' path in
-        ignore(Includemod.modtypes  ~mark:true ~loc sig_env
-                 newmd.md_type md.md_type);
-        return
-          ~replace_by:(Some(Sig_module(id, pres, newmd, rs, priv)))
-          (Pident id, Built_TypedTree {
-              lid=lid; constr=(Twith_module (path, lid'))})
-    | Sig_module(id, _, md, _rs, _), [s], With_modsubst (lid',path,md')
-      when Ident.name id = s ->
-        let sig_env = Env.add_signature sg_for_env outer_sig_env in
-        let aliasable = not (Env.is_functor_arg path sig_env) in
-        ignore
-          (Includemod.strengthened_module_decl ~loc ~mark:true
-             ~aliasable sig_env md' path md);
-        real_ids := [Pident id];
-        return ~replace_by:None
-          (Pident id, Built_TypedTree {
-               lid=lid; constr=(Twith_modsubst (path, lid'))})
-
-    (* When the constraint affects a component of a submodule *)
-    | Sig_module(id, _, md, rs, priv) as current_item, s :: namelist, _
+    with Includemod.Error explanation ->
+      raise(Error(loc, initial_env, With_mismatch(lid.txt, explanation)))
```

Comment:
This `try ... with ...` is used at every recursive call instead of only once in the previous version.
As `merge_signature` is called only by `merge` outside of this definition it might be cleaner to move it there.